### DISCUSSION TITLE
Save the tasks in the DAO

### DISCUSSION THREAD

**Discussion by fthankac** [Netflix/conductor#2888] [Q&A]
During schedule task (**Method Name : scheduleTask**) in WorkflowExecutor (core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java) line number 1718 doesn't return the taskmodel after saving the tasks in the DAO. 

executionDAOFacade.createTasks(tasks);

Previously it used to return only saved tasks. The saved tasks are used for further processing. 

Now the input passed (i.e All the tasks) to createTasks are used for further processing. Also the local variable 'createdTasks' is not used anywhere.

Example

// Save the tasks in the DAO
            executionDAOFacade.createTasks(tasks);
            List<TaskModel> systemTasks =
                    tasks.stream()
                            .filter(task -> systemTaskRegistry.isSystemTask(task.getTaskType()))
                            .collect(Collectors.toList());

**Comment by apanicker-nflx** [comment]
>Also the local variable 'createdTasks' is not used anywhere.

This variable is unused and will be removed in the next refactor.

>Previously it used to return only saved tasks. The saved tasks are used for further processing.

The implicit contract here was that all tasks need to be saved. So essentially all the tasks passed into createTasks method must be saved in the DAO. The return value did not impact execution/persistence and has been removed since.

**Reply by fthankac** [reply]
In redis-persistence/src/main/java/com/netflix/conductor/redis/dao/RedisExecutionDAO.java,  in the method "createTasks" , there is check added below 

Long added =
                    jedisProxy.hset(
                            nsKey(SCHEDULED_TASKS, task.getWorkflowInstanceId()),
                            taskKey,
                            task.getTaskId());
            if (added < 1) {
                LOGGER.debug(
                        "Task already scheduled, skipping the run "
                                + task.getTaskId()
                                + ", ref="
                                + task.getReferenceTaskName()
                                + ", key="
                                + taskKey);
                continue;
            }

These task will not get saved in the DAO when the task model has such type of tasks.


Because of the above condition, in core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java, line number 1073 throws error "No such task found by id:" when task is not saved in DAO.
